### PR TITLE
Clean up MC3 code + add some MC3 mixing statistics

### PR DESCRIPTION
**Pull request by bredelings**
This patch 
* cleans up the MC3 code
* adds mixing statistics that supplement the probability of swapping chains
* adds a comment and paper reference about the new mixing statistics.

There do not appear to be any MC3 tests in the testsuite, but I modified `test_JC_unrooted` (which is fast) to do the following:
```
mymcmc = mcmcmc(mymodel, monitors, moves, nchain=4, nruns=2, tuneHeat = TRUE)

mymcmc.burnin(generations=200,tuningInterval=20)
mymcmc.run(generations=2000)
```
According to my tests, there is no change to the output file `output{,4}/mcmcmc_JC_unrooted_uniform_run_1.out`.

### CODE REVIEW

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (jsigao) [inline src/core/analysis/mcmc/Mcmcmc.cpp:622]: Not entirely sure if heat_ranks needs to be synchronized (explicitly with a specific synchronization function) across processes, like the chain_heats, when MPI is used (seems unnecessary to me, but I never fully grasped the reason that the heats need to be explicitly synchronized...).

**Review by jsigao — COMMENTED** [review]
Hi Ben @bredelings -- thanks for cleaning up the code and adding the statistics! I looked through the changes and didn't spot any issue. I just have a couple rather superficial comments left below.

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (bredelings) [reply src/core/analysis/mcmc/Mcmcmc.cpp:622]: Good point.  Relying on heat_ranks[] here might not be correct, since the heat_ranks[] might be out of date with respect to the chain_heats[].

It is problematic that the same information is stored in multiple places -- they can get out of sync.  I started doing additional refactoring to 
* compute chain_heat(i) from heat_temps[] and heat_ranks[]
* synchronize heat_temps[] and heat_ranks[]
However, I decided to avoid doing such invasive changes in the first round.  The code could really use a second round of refactoring.

Comment (bredelings) [reply src/core/analysis/mcmc/Mcmcmc.cpp:622]: OK, this should be fixed now.

**Comment by bredelings** [comment]
@jsigao Does this seem OK now?

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (hoehna) [inline src/core/analysis/mcmc/Mcmcmc.cpp:69]: Why use "_" to separate word? We usually use  camel case

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (hoehna) [inline src/core/analysis/mcmc/Mcmcmc.cpp:1116]: I'm really not sure that this style of coding is the most readable. But maybe I'm just getting out of touch...

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (hoehna) [inline src/core/analysis/mcmc/Mcmcmc.cpp:1763]: Is this MPI safe? Not sure if I caught the details, but we often runs MC3 with MPI and use exactly as many cores as chains. So was this tested?

**Review by hoehna — CHANGES_REQUESTED** [review]

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (bredelings) [reply src/core/analysis/mcmc/Mcmcmc.cpp:1763]: I constructed an MC3 example and ran it under MPI with 1, 2, and 4 cores.  Given a fixed number of chains, the number of cores does not change the output.  I found and avoided a few mistakes that way.

I don't think any of the current tests use MC3, but we should probably add one that uses it.

Regarding the `std::swap( )` code, it does exact the same thing as the previous code.  It is just shorter.

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (bredelings) [reply src/core/analysis/mcmc/Mcmcmc.cpp]: I switched the new functions to use camel case.

File: src/core/analysis/mcmc/Mcmcmc.cpp
Comment (bredelings) [reply src/core/analysis/mcmc/Mcmcmc.cpp:1116]: Lots of types have `ostream& operator<<(ostream&, const T&)` defined.
All of them would work here.
This seems fine...

**Comment by bredelings** [comment]
@hoehna Thanks for the review!
I changed the new functions to use camelCase, and I have definitely tested the changes with MPI.
More MPI testing would be appreciated though!

**Comment by jsigao** [comment]
I think this all look good to me now! I'm ready to approve it as long as it's also okay with Sebastian @hoehna.

**Comment by bredelings** [comment]
@jsigao Sebastian has requested changes, so it can't be merged until he approves it as well.  He is on vacation this week, so hopefully next week.

**Review by jsigao — APPROVED** [review]
@bredelings Got it. (Sorry I haven't done many PR reviews before, so still not super familiar with the workflow...) Just approved.

**Review by hoehna — APPROVED** [review]
Great, looks good now.

**Pull request merged**
